File: pythonLeetcode/101-对称二叉树.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
给定一个二叉树，检查它是否是镜像对称的。

例如，二叉树 [1,2,2,3,4,4,3] 是对称的。

    1
   / \
  2   2
 / \ / \
3  4 4  3
但是下面这个 [1,2,2,null,3,null,3] 则不是镜像对称的:

    1
   / \
  2   2
   \   \
   3    3
说明:

如果你可以运用递归和迭代两种方法解决这个问题，会很加分。

来源：力扣（LeetCode）
链接：https://leetcode-cn.com/problems/symmetric-tree
著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。
"""

class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

def someTreeNode(lTree,rTree):
    if not lTree and rTree:
        return False
    elif lTree and not rTree:
        return False
    elif not lTree and not rTree:
        return True
    sameRoot = lTree.val == rTree.val
    if not sameRoot:
        return False
    lSame = someTreeNode(lTree.left,rTree.right)
    if not lSame:
        return False
    rSame = someTreeNode(lTree.right,rTree.left)
    if not rSame:
        return False
    return True

class Solution(object):
    def isSymmetric(self, root):
        if not root:
            return False
        return someTreeNode(root.left,root.right)
    

File: LoveSwiftLeetcode/101.对称二叉树.swift
/*
 * @lc app=leetcode.cn id=101 lang=swift
 *
 * [101] 对称二叉树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init() { self.val = 0; self.left = nil; self.right = nil; }
 *     public init(_ val: Int) { self.val = val; self.left = nil; self.right = nil; }
 *     public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
 *         self.val = val
 *         self.left = left
 *         self.right = right
 *     }
 * }
 */
class Solution {
    func leftToRightSymmetric(_ left: TreeNode?, _ right: TreeNode?) -> Bool {
        if left == nil , right != nil {
            return false
        } else if left != nil , right == nil {
            return false
        } else if left == nil , right == nil {
            return true
        }
        if left!.val != right!.val {
            return false
        }
        let rl = leftToRightSymmetric(left?.right, right?.left)
        if rl == false {
            return false
        }
        let lr = leftToRightSymmetric(left?.left, right?.right)
        if lr == false {
            return false
        }
        return true
    }
    func isSymmetric(_ root: TreeNode?) -> Bool {
        if root == nil {
            return false
        }
        return leftToRightSymmetric(root?.left, root?.right)
    }
}
// @lc code=end

